File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SavedThreadLoaderManager.java
package com.github.adamantcheese.chan.core.manager;

import android.net.Uri;

import androidx.annotation.Nullable;

import com.github.adamantcheese.chan.core.database.DatabaseManager;
import com.github.adamantcheese.chan.core.mapper.ThreadMapper;
import com.github.adamantcheese.chan.core.model.ChanThread;
import com.github.adamantcheese.chan.core.model.orm.Loadable;
import com.github.adamantcheese.chan.core.model.save.SerializableThread;
import com.github.adamantcheese.chan.core.repository.SavedThreadLoaderRepository;
import com.github.adamantcheese.chan.core.saf.FileManager;
import com.github.adamantcheese.chan.core.saf.file.ExternalFile;
import com.github.adamantcheese.chan.core.settings.ChanSettings;
import com.github.adamantcheese.chan.utils.BackgroundUtils;
import com.github.adamantcheese.chan.utils.Logger;
import com.google.gson.Gson;

import java.io.File;
import java.io.IOException;

import javax.inject.Inject;

public class SavedThreadLoaderManager {
    private final static String TAG = "SavedThreadLoaderManager";

    private final Gson gson;
    private final DatabaseManager databaseManager;
    private final SavedThreadLoaderRepository savedThreadLoaderRepository;
    private final FileManager fileManager;

    @Inject
    public SavedThreadLoaderManager(
            Gson gson,
            DatabaseManager databaseManager,
            SavedThreadLoaderRepository savedThreadLoaderRepository,
            FileManager fileManager) {
        this.gson = gson;
        this.databaseManager = databaseManager;
        this.savedThreadLoaderRepository = savedThreadLoaderRepository;
        this.fileManager = fileManager;
    }

    @Nullable
    public ChanThread loadSavedThread(Loadable loadable) {
        if (BackgroundUtils.isMainThread()) {
            throw new RuntimeException("Cannot be executed on the main thread!");
        }

        if (ChanSettings.localThreadsLocationUri.get().isEmpty()) {
            throw new IllegalStateException("Local threads location is not set!");
        }

        String threadSubDir = ThreadSaveManager.getThreadSubDir(loadable);
        Uri localThreadsLocationUri = Uri.parse(ChanSettings.localThreadsLocationUri.get());
        ExternalFile threadSaveDir = fileManager.fromUri(localThreadsLocationUri)
                .appendSubDirSegment(threadSubDir);

        if (!threadSaveDir.exists() || !threadSaveDir.isDirectory()) {
            Logger.e(TAG, "threadSaveDir does not exist or is not a directory: "
                    + "(path = " + threadSaveDir.getFullPath()
                    + ", exists = " + threadSaveDir.exists()
                    + ", isDir = " + threadSaveDir.isDirectory() + ")");
            return null;
        }

        ExternalFile threadFile = threadSaveDir
                .clone()
                .appendFileNameSegment(SavedThreadLoaderRepository.THREAD_FILE_NAME);

        if (!threadFile.exists() || !threadFile.isFile() || !threadFile.canRead()) {
            Logger.e(TAG, "threadFile does not exist or not a file or cannot be read: " +
                    "(path = " + threadFile.getFullPath()
                    + ", exists = " + threadFile.exists()
                    + ", isFile = " + threadFile.isFile()
                    + ", canRead = " + threadFile.canRead() + ")");
            return null;
        }

        ExternalFile threadSaveDirImages = threadSaveDir
                .clone()
                .appendSubDirSegment("images");

        if (!threadSaveDirImages.exists() || !threadSaveDirImages.isDirectory()) {
            Logger.e(TAG, "threadSaveDirImages does not exist or is not a directory: "
                    + "(path = " + threadSaveDirImages.getFullPath()
                    + ", exists = " + threadSaveDirImages.exists()
                    + ", isDir = " + threadSaveDirImages.isDirectory() + ")");
            return null;
        }

        try {
            SerializableThread serializableThread = savedThreadLoaderRepository
                    .loadOldThreadFromJsonFile(threadSaveDir);
            if (serializableThread == null) {
                Logger.e(TAG, "Could not load thread from json");
                return null;
            }

            return ThreadMapper.fromSerializedThread(loadable, serializableThread);
        } catch (IOException | SavedThreadLoaderRepository.OldThreadTakesTooMuchSpace e) {
            Logger.e(TAG, "Could not load saved thread", e);
            return null;
        }
    }

}

File: Kuroba/app/src/main/java/com/github/adamantcheese/chan/core/manager/SavedThreadLoaderManager.kt
package com.github.adamantcheese.chan.core.manager

import com.github.adamantcheese.chan.core.mapper.ThreadMapper
import com.github.adamantcheese.chan.core.model.ChanThread
import com.github.adamantcheese.chan.core.model.orm.Loadable
import com.github.adamantcheese.chan.core.repository.SavedThreadLoaderRepository
import com.github.adamantcheese.chan.ui.settings.base_directory.LocalThreadsBaseDirectory
import com.github.adamantcheese.chan.utils.BackgroundUtils
import com.github.adamantcheese.chan.utils.Logger
import com.github.k1rakishou.fsaf.FileManager
import com.github.k1rakishou.fsaf.file.DirectorySegment
import com.github.k1rakishou.fsaf.file.FileSegment
import java.io.IOException
import javax.inject.Inject

class SavedThreadLoaderManager @Inject constructor(
        private val savedThreadLoaderRepository: SavedThreadLoaderRepository,
        private val fileManager: FileManager
) {

    fun loadSavedThread(loadable: Loadable): ChanThread? {
        if (BackgroundUtils.isMainThread()) {
            throw RuntimeException("Cannot be executed on the main thread!")
        }

        val baseDir = fileManager.newBaseDirectoryFile<LocalThreadsBaseDirectory>()
        if (baseDir == null) {
            Logger.e(TAG, "loadSavedThread() fileManager.newLocalThreadFile() returned null")
            return null
        }

        val threadSaveDir = baseDir
                .clone(ThreadSaveManager.getThreadSubDir(loadable))

        val threadSaveDirExists = fileManager.exists(threadSaveDir)
        val threadSaveDirIsDirectory = fileManager.isDirectory(threadSaveDir)

        if (!threadSaveDirExists || !threadSaveDirIsDirectory) {
            Logger.e(TAG, "threadSaveDir does not exist or is not a directory: "
                    + "(path = " + threadSaveDir.getFullPath()
                    + ", exists = " + threadSaveDirExists
                    + ", isDir = " + threadSaveDirIsDirectory + ")")
            return null
        }

        val threadFile = threadSaveDir
                .clone(FileSegment(SavedThreadLoaderRepository.THREAD_FILE_NAME))

        val threadFileExists = fileManager.exists(threadFile)
        val threadFileIsFile = fileManager.isFile(threadFile)
        val threadFileCanRead = fileManager.canRead(threadFile)

        if (!threadFileExists || !threadFileIsFile || !threadFileCanRead) {
            Logger.e(TAG, "threadFile does not exist or not a file or cannot be read: " +
                    "(path = " + threadFile.getFullPath()
                    + ", exists = " + threadFileExists
                    + ", isFile = " + threadFileIsFile
                    + ", canRead = " + threadFileCanRead + ")")
            return null
        }

        val threadSaveDirImages = threadSaveDir
                .clone(DirectorySegment("images"))

        val imagesDirExists = fileManager.exists(threadSaveDirImages)
        val imagesDirIsDir = fileManager.isDirectory(threadSaveDirImages)

        if (!imagesDirExists || !imagesDirIsDir) {
            Logger.e(TAG, "threadSaveDirImages does not exist or is not a directory: "
                    + "(path = " + threadSaveDirImages.getFullPath()
                    + ", exists = " + imagesDirExists
                    + ", isDir = " + imagesDirIsDir + ")")
            return null
        }

        try {
            val serializableThread = savedThreadLoaderRepository
                    .loadOldThreadFromJsonFile(threadSaveDir)
            if (serializableThread == null) {
                Logger.e(TAG, "Could not load thread from json")
                return null
            }

            return ThreadMapper.fromSerializedThread(loadable, serializableThread)
        } catch (e: IOException) {
            Logger.e(TAG, "Could not load saved thread", e)
            return null
        }
    }

    companion object {
        private const val TAG = "SavedThreadLoaderManager"
    }

}
